### PR TITLE
FlowDeathPact

### PR DESCRIPTION
**Pull request by ktoso**
I implemented this as a response to the mailing list, but actually thought it might be very useful in general, WDYT?

### CODE REVIEW

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowDeathPactSpec.scala [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    override def preStart(): Unit = {
+      getStageActor({
+        case (_, Terminated(`ref`)) ⇒ onTerminated(this)
+        case _                      ⇒ // ignore...
+      })
+      stageActor.watch(ref)
+      ref.tell(FlowDeathPact.Signed, stageActor.ref)
+    }
+
+    override def onPush(): Unit = push(out, grab(in))
```
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowDeathPactSpec.scala]: random optimisation idea to somehow allow signalling "this stage actually does not care about the elements at all, just pass through please"

Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowDeathPactSpec.scala]: that is called rewiring and I worked on it but had no time to implement it. Next time ;)

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowDeathPactSpec.scala [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+      info("Stage sender: " + contrActor.lastSender)
+      probe.expectSubscriptionAndComplete()
+    }
+
+  }
+
+}
+
+/** INTERNAL API */
+/*private[akka]*/ final class FlowDeathPact[T](ref: ActorRef, onTerminated: GraphStageLogic ⇒ Unit)
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowDeathPactSpec.scala]: Make it FlowDeathWatch instead?

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowDeathPactSpec.scala [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  override def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) 
+    with InHandler with OutHandler {
+
+    override def preStart(): Unit = {
+      getStageActor({
+        case (_, Terminated(`ref`)) ⇒ onTerminated(this)
+        case _                      ⇒ // ignore...
+      })
+      stageActor.watch(ref)
+      ref.tell(FlowDeathPact.Signed, stageActor.ref)
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowDeathPactSpec.scala]: is this needed?

**Comment by drewhk** [comment]
This feels contrib to me. I feel in its current form this is rather specialized. 

What I can imagine being more useful (as in more basic and composable) is a stage that materializes to an ActorRef that can be watched and can be sent a PosionPill and ignores everything else.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowDeathPactSpec.scala (hunk 3 above)
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowDeathPactSpec.scala]: Thought might be useful if the actor maintains registry of flows it's watching and should terminate once all of the flows that registered with it have terminated

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowDeathPactSpec.scala (hunk 1 above)
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowDeathPactSpec.scala]: I see, excellent :)

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowDeathPactSpec.scala (hunk 2 above)
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowDeathPactSpec.scala]: Yeah, good name

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowDeathPactSpec.scala (hunk 3 above)
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowDeathPactSpec.scala]: For that, the ActorRef is enough. Plus, the built-in watch mechanism is much more reliable and works over the wire. Hence I think it is much better to expose the stage ActorRef and let the other side watch it if it wants to.

**Comment by patriknw** [comment]
In one direction we already have `watchTermination`. In the other it is possible to do it with KillSwitch, although you must spawn an actor to do the watch. Not sure about this one. I agree with @drewhk that it should start in contrib if anywhere.

**Comment by johanandren** [comment]
So, close this and move to contrib?

**Comment by ktoso** [comment]
I'm ok with moving to contrib.

**Pull request closed without merging**